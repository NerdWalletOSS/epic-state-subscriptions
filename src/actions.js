/**
 * @param {String} type - The name of the action
 * @return {Function} - The action creator function used to create redux actions
 */
export const makeActionCreator = type => {
  const actionWrapper = payload => ({
    type,
    payload
  });
  actionWrapper.type = type;
  return actionWrapper;
};

/**
 * The action used to dynamically override the subscriptions paths for a state subscription
 * @param {Object} payload - The payload of the action to be processed by the reducer
 * @param {String} payload.stateSubscriptionKey - The state subscription key whose paths are to be overridden
 * @param {Array} payload.paths - The paths with which the state subscription key should be overridden
 */
export const overrideStateSubscriptionPaths = makeActionCreator(
  "UPDATE_STATE_SUBSCRIPTION_PATHS"
);
