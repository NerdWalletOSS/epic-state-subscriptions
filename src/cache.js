import _ from "lodash";

// The state subscription path cache records cache entries for each distinct state subscription path the client has configured to watch.
// When the state in the cache is different from the current state of the Redux store, a cache miss has occurred
// and the client's path operator will receive the updated path in their path stream which they can then process.
export const stateSubscriptionPathCache = {};

export const getCachedPathsForSubscription = subscriptionKey => {
  const cachedPathPatterns = _.get(
    exports.stateSubscriptionPathCache,
    subscriptionKey
  );
  if (cachedPathPatterns) {
    return Object.keys(cachedPathPatterns).flatMap(cachedPathPatternKey => {
      return Object.keys(cachedPathPatterns[cachedPathPatternKey]).map(
        cachedPath => ({ path: cachedPath, pathPattern: cachedPathPatternKey })
      );
    });
  }
  return [];
};
