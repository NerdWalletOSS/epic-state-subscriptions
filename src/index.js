export { overrideStateSubscriptionPaths } from "./actions";
export { createStateSubscriptionEpics } from "./epics";
export { stateSubscriptionReducer } from "./reducers";
export { getStateSubscriptionOverridePaths } from "./selectors";
