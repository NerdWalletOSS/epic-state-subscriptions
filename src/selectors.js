import _ from "lodash";

/**
 * A selector method used to retrieve the overridden state subscription paths currently being used
 * for the given state subscription key
 * @param {Object} state - The redux store
 * @param {String} subscriptionKey - the state subscription key for which paths should be retrieved
 * @return {Array?} - The state subscription override paths for the given state subscription key
 */
export const getStateSubscriptionOverridePaths = (state, subscriptionKey) =>
  _.get(state, `stateSubscriptions.${subscriptionKey}.paths`);
