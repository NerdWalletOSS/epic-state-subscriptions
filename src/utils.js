import _ from "lodash";

/**
 * A basic shallow equal comparison utility
 * @param {Object?} - The first object to be shallow compared
 * @param {Object?} - The second object to be shallow compared
 * @return {Boolean} Whether the two objects are shallow equal, defaults to true for === equality
 * of objects and primitives
 */
export const shallowEqual = (obj1, obj2) => {
  if (!obj1 || !obj2) {
    return obj1 === obj2;
  }
  return _.every(
    [...Object.keys(obj1), ...Object.keys(obj2)],
    key => obj1[key] === obj2[key]
  );
};

/**
 * Determines if there are dirty state subscription paths for a nested path where nested is defined as a
 * `*` property such as `state.x.*`. It traverses each top level key off of `state.x` and aggregates dirty paths.
 *
 * @param {String} stateSubscriptionKey - The state subscription identifier key
 * @param {String} stateSubscriptionPathCache - The subscription cache used to compare the current redux state with the cached value
 * @param {Array} pathComponents - The state subscription path delimited into an array by `.`
 * @param {Object} state - The current state of the Redux store
 * @param {Number} pathIndex - The path component index currently being traversed
 * @param {Function} callback - The dirty subscription callback to be invoked on each transformed nested path
 * @return {Array} The dirty paths in the redux store for the provided state subscription path
 */
const findNestedDirtySubscriptionPaths = (
  stateSubscriptionKey,
  stateSubscriptionPathCache,
  pathPattern,
  pathComponents,
  state,
  pathIndex,
  callback
) => {
  const stringPath = pathComponents.slice(0, pathIndex).join(".");
  const stateSlice = pathIndex === 0 ? state : _.get(state, stringPath);
  // If the state of the redux store is an object, we want to traverse each key under the object
  // to determine if the top-level keys are dirty.
  if (_.isPlainObject(stateSlice)) {
    return Object.keys(stateSlice).reduce((acc, slicePath) => {
      // For each traversed top-level key, check if it is a dirty path by creating its path components
      // array by taking the preceding components, its own key name, then the path components after it.
      // Consider the case where `state.x.y.z` exists and we are recursing on the state subscription `state.x.*.z`
      // Here it would be looping over all the keys for `state.x`, discover y and then recurse with the path `state.x.y.z`
      const newPathComponents = [
        ...pathComponents.slice(0, pathIndex),
        slicePath,
        ...pathComponents.slice(pathIndex + 1, pathComponents.length)
      ];
      return [
        ...acc,
        ...callback(
          stateSubscriptionKey,
          stateSubscriptionPathCache,
          pathPattern,
          newPathComponents,
          state,
          pathIndex
        )
      ];
    }, []);
  }
  return [];
};

/**
 * Determines if there is a dirty state subscription path for the given path such as `state.x`.
 *
 * @param {String} stateSubscriptionKey - The state subscription identifier key
 * @param {String} stateSubscriptionPathCache - The subscription cache used to compare the current redux state with the cached value
 * @param {String} pathPattern - The original state subscription path pattern
 * @param {Array} pathComponents - The state subscription path delimited into an array by `.`
 * @param {Object} state - The current state of the Redux store
 * @param {Number} pathIndex=0 - The path component index currently being traversed
 * @param {Array} foundPaths=[] - The current dirty state subscription paths found
 * @return {Array} The dirty paths in the redux store for the provided state subscription path
 */
const findDirtySubscriptionPaths = (
  stateSubscriptionKey,
  stateSubscriptionPathCache,
  pathPattern,
  pathComponents,
  state,
  pathIndex = 0,
  foundPaths = []
) => {
  const pathComponent = pathComponents[pathIndex];
  // Nested Recursive case:
  // If the current state subscription path component being traversed is a `*`, we want to check all paths
  // under the current state, so we call the helper method `findNestedDirtySubscriptionPaths` to check
  // the nested paths under the current state subscription path.
  if (pathComponent === "*") {
    return [
      ...foundPaths,
      ...findNestedDirtySubscriptionPaths(
        stateSubscriptionKey,
        stateSubscriptionPathCache,
        pathPattern,
        pathComponents,
        state,
        pathIndex,
        findDirtySubscriptionPaths
      )
    ];
    // Base case:
    // If we are traversing the path components and have reached the final one, we want to check the state
    // at the full string path re-created by joining the preceding and current state subscription path components with
    // the state subscription cache for that path.
  } else if (pathIndex === pathComponents.length - 1) {
    const stringPath = pathComponents.join(".");

    // The cached paths are retrieved from {subscriptionKey}{pathPattern}{cacheHitKey}
    // For example, a top level key of `group1` with a path pattern of `test.*` and 2 keys `foo` and
    // `bar` under the test slice of the redux store would store the cached entry for each at paths
    // `group1.test.*.foo and `group1.test.*.bar` in the cache
    const stateSubscriptionPathCacheForPattern = _.get(
      stateSubscriptionPathCache[stateSubscriptionKey],
      pathPattern
    );
    const subscriptionPath = _.get(
      stateSubscriptionPathCacheForPattern,
      stringPath
    );
    const stateSlice = _.get(state, stringPath);
    if (!shallowEqual(stateSlice, subscriptionPath)) {
      // If the state subscription path is dirty, update the state subscription cache for that path.
      // The cached paths are stored under the pattern {subscriptionKey}{pathPattern}{cacheHitKey}
      _.set(
        stateSubscriptionPathCache,
        [stateSubscriptionKey, pathPattern, stringPath],
        stateSlice
      );
      return [...foundPaths, { pathPattern, path: stringPath }];
    }
    return foundPaths;
  }
  // Recursive case:
  // Only a partial path has been traversed so far so it continues the recursion at the next path index
  return findDirtySubscriptionPaths(
    stateSubscriptionKey,
    stateSubscriptionPathCache,
    pathPattern,
    pathComponents,
    state,
    pathIndex + 1,
    foundPaths
  );
};

/**
 * Checks the given state subscription key to determine if any of its watched paths
 * are dirty in the Redux store.
 *
 * @param {String} stateSubscriptionKey - The state subscription identifier key
 * @param {String} stateSubscriptionPathCache - The subscription cache used to compare the current redux state with the cached value
 * @param {Array} paths - The paths that the individual state subscription config is subscribed to
 * @param {Object} state - The current Redux state
 * @return {Array} - The dirty state subscription paths
 */
// eslint-disable-next-line import/prefer-default-export
export const findUpdatedStateSubscriptionPaths = (
  stateSubscriptionKey,
  stateSubscriptionPathCache,
  pathPatterns,
  state
) =>
  // Aggregate all the paths that have changed together
  pathPatterns.reduce(
    (acc, pathPattern) => [
      ...acc,
      ...findDirtySubscriptionPaths(
        stateSubscriptionKey,
        stateSubscriptionPathCache,
        pathPattern,
        pathPattern.split("."),
        state
      )
    ],
    []
  );
