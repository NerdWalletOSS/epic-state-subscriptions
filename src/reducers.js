import { overrideStateSubscriptionPaths } from "./actions";

export const initialState = {};

// The state subscription reducer that can be optionally included to dynamically override subscription
// paths using the overrideStateSubscriptionPaths action
export const stateSubscriptionReducer = (state = initialState, action = {}) => {
  switch (action.type) {
    case overrideStateSubscriptionPaths.type: {
      const {
        payload: { key, paths }
      } = action;
      return Object.assign({}, state, {
        [key]: paths
      });
    }
    default:
      return state;
  }
};
