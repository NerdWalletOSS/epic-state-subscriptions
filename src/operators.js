import _ from "lodash";
import { filter, map } from "rxjs/operators";
import uuid from "uuid/v4";
import { findUpdatedStateSubscriptionPaths } from "./utils";
import { getStateSubscriptionOverridePaths } from "./selectors";

// The state subscription path cache records cache entries for each distinct state subscription path the client has configured to watch.
// When the state in the cache is different from the current state of the Redux store, a cache miss has occurred
// and the client's path operator will receive the updated path in their path stream which they can then process.
const stateSubscriptionPathCache = {};

/**
 * An RxJS operator that converts a stream of actions into a stream of updated paths emitted
 * when a path monitored by the state subscription changes
 *
 * @param {Observable} state$ - The state stream from the epic that the operator requires
 * @param {Object} config - The config options for the state subscription operator
 * @param {String?} config.key - The key to use to distinguish each state subscription configuration for the purpose
 * of dynamic path overrides configurable via the `updateStateSubscriptionPaths` action
 * @param {Array?} config.paths - The paths that the client has configured to watch
 * initiated by the client
 * @return {Observable} - An observable that emits path change objects
 */
export const createStateSubscription = (state$, config) => action$ => {
  // default to an arbitrary uuid if the client does not specify a desired key for the state subscription.
  const { key: subscriptionKey = uuid(), paths: pathPatterns = [] } = config;

  // Initialize the cache for the state subscription so that an update in the Redux store to that path will
  // cause a cache miss. If it already exists then the operator has already been used with that key
  // and on re-creation it should not clear the existing cache.
  if (!stateSubscriptionPathCache[subscriptionKey]) {
    stateSubscriptionPathCache[subscriptionKey] = {};
  }

  // Modify the stream of actions received by the epic to return a stream of dirty string paths to be handled by the
  // state subscription path operator.
  const filteredStateSubscriptionPaths$ = action$.pipe(
    map(() => {
      const currentState = state$.value;
      const currentStateSubscriptionPathPatterns =
        getStateSubscriptionOverridePaths(state$.value, subscriptionKey) ||
        pathPatterns;

      // Filter the cache to only include entries for patterns that are still being subscribed to,
      // this way if the formerly subscribed path is re-added, there will be a cache miss
      const currentStateSubscriptionPathCache =
        stateSubscriptionPathCache[subscriptionKey];
      if (currentStateSubscriptionPathCache) {
        stateSubscriptionPathCache[subscriptionKey] = _.pick(
          currentStateSubscriptionPathCache,
          currentStateSubscriptionPathPatterns
        );
      }

      // Paths are collected in the format { path: path, pathPattern: pathPattern }
      return findUpdatedStateSubscriptionPaths(
        subscriptionKey,
        stateSubscriptionPathCache,
        currentStateSubscriptionPathPatterns,
        currentState
      );
    }),
    filter(paths => paths.length > 0)
  );

  // Returns the stream that was first processed into paths and then delivered to the client path operator
  // which executes client logic optionally triggering actions or performing side effects.
  return filteredStateSubscriptionPaths$;
};
