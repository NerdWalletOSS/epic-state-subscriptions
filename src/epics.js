import _ from "lodash";
import { filter, map } from "rxjs/operators";
import uuid from "uuid/v4";
import { getStateSubscriptionOverridePaths } from "./selectors";
import { findUpdatedStateSubscriptionPaths } from "./utils";

// The state subscription path cache records cache entries for each distinct state subscription path the client has configured to watch.
// When the state in the cache is different from the current state of the Redux store, a cache miss has occurred
// and the client's path operator will receive the updated path in their path stream which they can then process.
const stateSubscriptionPathCache = {};

/**
 * This function wraps the state subscription path operator from each state subcription config
 * within a state subscription epic that filters updates so that only a stream of paths that they have configured
 * to watch are sent to the path operators when the path's value changes in the Redux state.
 *
 * @param {String} stateSubscriptionEpicConfig.key - The key to use to distinguish each state subscription configuration for the purpose
 * of dynamic path overrides configurable via the `overrideStateSubscriptionPaths` action
 * @param {Array} stateSubscriptionEpicConfig.paths - The paths that the client has configured to watch
 * initiated by the client
 * @param {Function} stateSubscriptionEpicConfig.pathOperator - The path operator function which takes a stream of updated string paths and performs
 * the desired client behaviour
 * @return {Array} - Returns an array of state subscription epics that wrap the path operators provided by the client
 * configuration which receive the stream of updated path changes
 */
export const createStateSubscriptionEpics = stateSubscriptionConfigs =>
  stateSubscriptionConfigs.map(
    // Map the state subscription configuration provided by the client to a state subscription epic.
    stateSubscriptionConfig => (action$, state$) => {
      const {
        key,
        paths: initialSubscriptionPaths,
        pathOperator
      } = stateSubscriptionConfig;
      // default to an arbitrary uuid if the client does not specify a desired key for the state subscription.
      const subscriptionKey = key || uuid();
      // Initialize the cache for the state subscription so that an update in the Redux store to that path will
      // cause a cache miss.
      stateSubscriptionPathCache[subscriptionKey] = {};
      // Modify the stream of actions received by the epic to return a stream of dirty string paths to be handled by the
      // state subscription path operator.
      const filteredStateSubscriptionPaths$ = action$.pipe(
        map(() => {
          const currentState = state$.value;
          const currentStateSubscriptionPathPatterns =
            getStateSubscriptionOverridePaths(currentState, subscriptionKey) ||
            initialSubscriptionPaths;

          // Filter the cache to only include entries for patterns that are still being subscribed to,
          // this way if the formerly subscribed path is re-added, there will be a cache miss
          const currentStateSubscriptionPathCache =
            stateSubscriptionPathCache[subscriptionKey];
          if (currentStateSubscriptionPathCache) {
            stateSubscriptionPathCache[subscriptionKey] = _.pick(
              currentStateSubscriptionPathCache,
              currentStateSubscriptionPathPatterns
            );
          }

          // Paths are collected in the format { path: path, pathPattern: pathPattern }
          return findUpdatedStateSubscriptionPaths(
            subscriptionKey,
            stateSubscriptionPathCache,
            currentStateSubscriptionPathPatterns,
            currentState
          );
        }),
        filter(paths => paths.length > 0)
      );
      // Returns the stream that was first processed into paths and then delivered to the client path operator
      // which executes client logic optionally triggering actions or performing side effects.
      return pathOperator(filteredStateSubscriptionPaths$, state$);
    }
  );
