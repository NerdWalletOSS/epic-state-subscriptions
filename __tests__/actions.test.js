import { initialState, stateSubscriptionReducer } from "../src/reducers";
import { overrideStateSubscriptionPaths } from "../src/actions";

describe("stateSubscriptionReducer", () => {
  describe("with no actions", () => {
    test("should have expected initial state", () => {
      expect(stateSubscriptionReducer(initialState)).toMatchSnapshot();
    });
  });

  describe("with an overrideStateSubscriptionPaths action", () => {
    test("should override the state subscription entry with the given paths", () => {
      expect(
        stateSubscriptionReducer(
          initialState,
          overrideStateSubscriptionPaths({
            key: "subscriptionKey",
            paths: ["testPath"]
          })
        )
      ).toMatchSnapshot();
    });
  });
});
