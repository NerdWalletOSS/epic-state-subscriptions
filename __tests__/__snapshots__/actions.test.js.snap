// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stateSubscriptionReducer with an overrideStateSubscriptionPaths action should override the state subscription entry with the given paths 1`] = `
Object {
  "subscriptionKey": Array [
    "testPath",
  ],
}
`;

exports[`stateSubscriptionReducer with no actions should have expected initial state 1`] = `Object {}`;
